### PR TITLE
create mrz correction

### PR DESCRIPTION
**Pull request by wadjih-bencheikh18**
closes https://github.com/cheminfo/mrz/issues/22

### CODE REVIEW

File: src/formats.ts [hunk 1]
```diff
@@ -1,12 +1,15 @@
...
 };
-Object.freeze(formats);
-
-export default formats;
+export type FormatType =
+  | 'TD1'
+  | 'TD2'
+  | 'TD3'
+  | 'SWISS_DRIVING_LICENSE'
+  | 'FRENCH_NATIONAL_ID';
```
[the remark this answers is not in the document]
Comment (wadjih-bencheikh18) [reply src/formats.ts:15]: I could implement it because formats.TD1 now is not FormatType

[the remark this answers is not in the document]
Comment (targos) [reply src/formats.ts:15]: does it work if you do `const formats = { ... } as const` ?

[the remark this answers is not in the document]
Comment (wadjih-bencheikh18) [reply src/formats.ts:15]: thank you

**Comment by lpatiny** [comment]
@targos In the issue I asked:

> For all the fields (and likely also at the level of fieldTemplates) we will probably need a property that specifies the allowed values (NUMERIC, CHARACTERS, BOTH)

But it was implemented as separate files like 'td1Correction'. I think this implementation is more difficult to maintain but maybe I missed something.

**Comment by targos** [comment]
I agree.

**Comment by wadjih-bencheikh18** [comment]
I refactored the code (correct field before parsing), but there is one issue in the autocorrect
there's some fields that have a related ranges, the parse function is using it without correction

I see that the solution is to correct all the line then start to parse fields

File: src/parse/createFieldParser.ts [hunk 2]
```diff
@@ -20,21 +24,32 @@ interface ParseResult {
 
 type Parser = (source: string, ...related: string[]) => ParseResult | string;
 
-type FieldOptions = {
+type FieldTypes = keyof typeof fieldTypes;
+export const fieldTypes = {
+  NUMERIC: 'NUMERIC',
+  CHARACTERS: 'CHARACTERS',
+  BOTH: 'BOTH',
```
Comment (stropitek) [inline src/parse/createFieldParser.ts:31]: I think you can name that "alphanumeric"

File: src/parse/createFieldParser.ts [hunk 3]
```diff
@@ -20,21 +24,32 @@ interface ParseResult {
 
 type Parser = (source: string, ...related: string[]) => ParseResult | string;
 
-type FieldOptions = {
+type FieldTypes = keyof typeof fieldTypes;
+export const fieldTypes = {
+  NUMERIC: 'NUMERIC',
+  CHARACTERS: 'CHARACTERS',
```
Comment (stropitek) [inline src/parse/createFieldParser.ts:30]: This can be named "alphabetic"

**Comment by targos** [comment]
> I see that the solution is to correct all the line then start to parse fields

Sounds good to me. A first loop could create the initial "empty" detail object for each field and do the correction. Then another pass is done to parse.

**Comment by wadjih-bencheikh18** [comment]
> Sounds good to me.

Implemented.

File: src/parse/utils/getInterval.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
+'use strict';
```
Comment (targos) [inline src/parse/utils/getInterval.ts:1]: I think this file isn't used anymore? You can delete it and the test.

**Pull request merged**